### PR TITLE
Added dedicated page about using types with pytest #12842

### PR DESCRIPTION
**Pull request by mwychung**
Fix (#12842 )

* *Added dedicated page about using types with pytest, `doc/en/how-to/types.rst`.*
This guide explains how to apply typing in pytest, covering typing test functions, fixtures, and parameterized tests to improve code readability and prevent type-related errors.

- [X] Allow maintainers to push and squash when merging my commits. 
- [X] Create a new changelog file.
- [X] Updated `AUTHORS`.

### CODE REVIEW

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you for your work on this. I feel like this not something that is specific to pytest so probably best as an external article or tutorial (?)

**Comment by mwychung** [comment]
Thank you for the feedback! I’m still a beginner with pytest and open-source contributions, and I’m eager to learn more. I appreciate your guidance. I’d be happy to focus on pytest-specific examples for typing if that would be more relevant. Would you like me to demonstrate typing with pytest’s fixtures, parameterized tests, or any specific areas you think would be most beneficial for the documentation?

Thanks again for your help!

**Review by nicoddemus — CHANGES_REQUESTED** [review]
@Pierre-Sassoulas actually this has been discussed in the past, and seems like the overall consensus is that having a page dedicated how to type tests would be welcome, as those topics are often brought up and unclear for many people.

However I think the docs as presented are a bit verbose, IMHO.

I think a "typing tests with pytest" page should already consider that readers know about Python typing (how to add types, what are the benefits/drawbacks, etc, perhaps only pointing to the external docs for readers that are not familiar yet), and focus mainly on the topics that pertain to pytest/tests: tests, fixtures, parametrized tests, etc.

The main benefits of add typing *to tests* should also be highlighted:

1. Refactoring: the type checker can easily point out the necessary changes in both code and tests, without needing to run the full test suite)
2. Readability: the types make the test code easier to read.

For production code, there are other benefits (such as preventing errors), but for tests, those are the main two reasons to add typing to tests IMHO.

File: doc/en/how-to/types.rst [hunk 1]
```diff
@@ -0,0 +1,142 @@
+.. _types:
+
+How to use Types with Pytest
+===========================================
+
+You can add typing in pytest functions that helps to specify what type of data a function expects and returns.
+This guide explains how to apply typing in pytest to improve code readability and prevent type-related errors.
+
+
+1. Introduction to Typing
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:10]: I would replace this entire section with a paragraph like "this page assumes the reader already knows how typing in Python works, its benefits, etc, for more information see PAGE".

File: doc/en/how-to/types.rst [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        return x + y
+
+In this example:
+`x: int` and `y: int` mean that `x` and `y` should be integers.
+`-> int` shows that the function returns an integer.
+
+It makes the intentions clear that both the input and output are expected to be integers.
+
+2. Typing Test Functions
+-----------------
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:30]: Sections need to use the same number of characters as the title, here for example:

```suggestion
2. Typing Test Functions
------------------------
```

File: doc/en/how-to/types.rst [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+Test functions in pytest check whether the code runs correctly.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:31]: As I mentioned, the reason for a test is different: it is better to run the test in order to see if the code is correct -- but we add typing to help with refactoring and readability. 👍

File: doc/en/how-to/types.rst [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    def test_add() -> None:
+        result = add(2, 3)
+        assert result == 5
+
+In this function:
+`test_add` is typed as `-> None` because it does not return anything.
+The assertion `assert result == 5` checks if the result is correct.
+
+Example result:
+If the input data type is incorrect, like passing `add("2", 3)`, a `TypeError` will occur.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:50]: These are not needed, as we assume readers already know about Python typing.

File: doc/en/how-to/types.rst [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+In this example:
+`sample_fixture()` is typed to return an `int`.
+In `test_sample_fixture`, using typing with fixtures, it ensures the return is an integer.
+If you change the return from an integer to a string (``"42"``), the test will fail.
+
+
+* Typing Fixtures with Lists and Dictionaries:
+Here are the examples showing how to use List and Dict types in pytest.
+When you want to use complex data structures like lists or dictionaries, import `List` and `Dict` from Python's `typing` module to specify the types.
+Note: From Python 3.5 or later, typing module is built-in module in Python.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:83]: We can drop this too, again we can assume the reader knows about typing and the particulars about `list`/`List`, etc.

File: doc/en/how-to/types.rst [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    import pytest
+
+
+    def test_add() -> None:
+        result = add(2, 3)
+        assert result == 5
+
+In this function:
+`test_add` is typed as `-> None` because it does not return anything.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:46]: We can instead mention that while at first marking a function as returning `-> None` might seem useless, having at least a type annotation means most type checkers will now type check the function, so it is useful because of that.

**Review by nicoddemus — CHANGES_REQUESTED** [review]

File: doc/en/how-to/types.rst (hunk 1 above)
Comment (mwychung) [reply doc/en/how-to/types.rst]: Thanks for the feedback! I added "Why Type tests" to focus on the specific benefits of typing in tests, like improved refactoring and readability. Let me know if further changes are needed!

File: doc/en/how-to/types.rst (hunk 2 above)
Comment (mwychung) [reply doc/en/how-to/types.rst]: Thank you so much for your helpful guidance. I matched all the section titles with same number of characters.

**Comment by mwychung** [comment]
Thank you so much for all the helpful feedback and guidance! I’ve removed the unnecessary sections as suggested and have added a few more focused areas specific to pytest, such as typing with monkeypatch, parametrize, and tmp_path. I believe these changes better align with the needs of the readers and improve the relevance of the documentation. If any further adjustments are needed, please feel free to let me know—I’m happy to make additional updates. I truly appreciate your insights throughout this process, and I’m excited to continue learning and contributing!

File: doc/en/how-to/types.rst [hunk 7]
```diff
@@ -0,0 +1,236 @@
+.. _types:
+
+Enhancing Type Annotations with Pytest
+======================================
+
+This page assumes the reader is familiar with Python's typing system and its advantages.
+For more information, refer to `Python's Typing Documentation <https://docs.python.org/3/library/typing.html>`_.
+
+Why Type Tests?
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:9]:
```suggestion
Why type tests?
```

File: doc/en/how-to/types.rst [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+    def test_add() -> None:
+        result = add(2, 3)
+        assert result == 5
+
+Here, `test_add` is annotated with `-> None`, as it does not return a value.
+While `-> None` typing may seem unnecessary, it ensures type checkers validate the function and helps identifying potential issues during refactoring.
+
+
+Typing Fixtures
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:47]:
```suggestion
Typing fixtures
```

File: doc/en/how-to/types.rst [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+Here:
+
+- **tmp_path_factory: pytest.TempPathFactory:** Indicates that `tmp_path_factory` is an instance of pytest’s `TempPathFactory`, responsible for creating temporary directories and paths during testing.
+- **fn: Path:** Identifies that `fn` is a `Path` object, emphasizing its role as a file path and clarifying the expected file operations.
+- **Return type -> Path:** Specifies the fixture returns a `Path` object, clarifying its expected structure.
+- **image_file: Path:** Defines `image_file` as a Path object, ensuring compatibility with `load_image`.
+
+Conclusion
+----------
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:234]:
```suggestion
----------

```

File: doc/en/how-to/types.rst [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+Typing tests in pytest provide unique advantages distinct from typing production code. Typed tests emphasize robustness in edge cases and diverse datasets.
+Type annotations provide an additional layer of validation, reducing the risk of runtime failures.
+
+- **Test Clarity:** Clearly defines expected inputs and outputs, improving readability, especially in complex or parameterized tests.
+
+- **Type Safety:** Helps catch mistakes in test data early, reducing runtime errors.
+
+- **Refactoring Support:** Serves as in-code documentation, clarifying data expectations and minimizing errors during test suite modifications.
+
+These benefits make typed tests a powerful tool for maintaining clarity, consistency, and safety throughout the testing process.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:21]:
```suggestion
Typing tests provides significant advantages: 

- **Readability:** Clearly defines expected inputs and outputs, improving readability, especially in complex or parameterized tests.

- **Refactoring:** This is the main benefit in typing tests, as it will greatly help with refactoring, letting the type checker point out the necessary changes in both production and tests, without needing to run the full test suite.

For production code, typing also helps catching some bugs that might not be caught by tests at all (regardless of coverage), for example:

.. code-block:: python

    def get_caption(target: int, items: list[tuple[int, str]]) -> str:
        for value, caption in items:
            if value == target:
                return caption
                
                    
The type checker will correctly error out that the function might return `None`, however even a full coverage test suite might miss that case:

.. code-block:: python

    def test_get_caption() -> None:
        assert get_caption(10, [(1, "foo"), (10, "bar")]) == "bar"


Note the code above has 100% coverage, but the bug is not caught (of course the example is "obvious", but serves to illustrate the point).
```

File: doc/en/how-to/types.rst [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    def add(a: int, b: int) -> int:
+        return a + b
+
+
+    def test_add() -> None:
+        result = add(2, 3)
+        assert result == 5
+
+Here, `test_add` is annotated with `-> None`, as it does not return a value.
+While `-> None` typing may seem unnecessary, it ensures type checkers validate the function and helps identifying potential issues during refactoring.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:44]: This entire section is redundant with the one above.

```suggestion
```

File: doc/en/how-to/types.rst [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    @pytest.fixture
+    def sample_dict() -> Dict[str, int]:
+        return {"a": 50, "b": 100}
+
+
+    def test_sample_dict(sample_dict: Dict[str, int]) -> None:
+        assert sample_dict["a"] == 50
+
+Annotating fixtures with types like List[int] and Dict[str, int] ensures data consistency and helps prevent runtime errors when performing operations.
+This ensures that only `int` values are allowed in the list and that `str` keys map to `int` values in the dictionary, helping avoid type-related issues.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:97]:
```suggestion

To type fixtures in pytest, just add normal types to the fixture functions -- there is nothing special that needs to be done just because of the `fixture` decorator.

.. code-block:: python

    import pytest


    @pytest.fixture
    def sample_fixture() -> int:
        return 38
        
In the same manner, the fixtures passed to test functions need be annotated with the fixture's return type:

.. code-block:: python

    def test_sample_fixture(sample_fixture: int) -> None:
        assert sample_fixture == 38

From the POV of the type checker, it does not matter that `sample_fixture` is actually a fixture managed by pytest, all it matters to it is that `sample_fixture` is a parameter of type `int`.
```

File: doc/en/how-to/types.rst [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+    import pytest
+
+
+    @pytest.mark.parametrize("input_value, expected_output", [(1, 2), (5, 6), (10, 11)])
+    def test_increment(input_value: int, expected_output: int) -> None:
+        assert input_value + 1 == expected_output
+
+Here, typing clarifies that both `input_value` and `expected_output` are expected as integers, promoting consistency.
+While parameterized tests can involve varied data types and that annotations simplify maintenance when datasets grow.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:115]:
```suggestion

The same logic applies to `@pytest.mark.parametrize`:

.. code-block:: python

    import pytest

    @pytest.mark.parametrize("input_value, expected_output", [(1, 2), (5, 6), (10, 11)])
    def test_increment(input_value: int, expected_output: int) -> None:
        assert input_value + 1 == expected_output

```

File: doc/en/how-to/types.rst [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+Here:
+
+- **username: Optional[str]:** Indicates the variable `username` may either be a string or `None`.
+- **get_os_user_lower() -> str:** Specifies this function will return a string, providing explicit return value type.
+- **monkeypatch fixture is typed as pytest.MonkeyPatch:** Shows that it will provide an object for patching environment variables during the test. This clarifies the intended use of the fixture and helps developers to use it correctly.
+- **Fixture return ->  None, like mock_env_user:** Specifies they do not return any value, but instead modify the test environment.
+
+Typing annotations can also be extended to `monkeypatch` usage in pytest for class methods, instance attributes, or standalone functions.
+This enhances type safety and clarity when patching the test environment.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:172]:
```suggestion

The same logic applies when typing fixture functions which receive other fixtures:

.. code-block:: python

    @pytest.fixture
    def mock_env_user(monkeypatch: pytest.MonkeyPatch) -> None:
        monkeypatch.setenv("USER", "TestingUser")

```

File: doc/en/how-to/types.rst [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+    def test_histogram(image_file: Path) -> None:
+        img = load_image(image_file)
+        # compute and test histogram
+
+Here:
+
+- **tmp_path_factory: pytest.TempPathFactory:** Indicates that `tmp_path_factory` is an instance of pytest’s `TempPathFactory`, responsible for creating temporary directories and paths during testing.
+- **fn: Path:** Identifies that `fn` is a `Path` object, emphasizing its role as a file path and clarifying the expected file operations.
+- **Return type -> Path:** Specifies the fixture returns a `Path` object, clarifying its expected structure.
+- **image_file: Path:** Defines `image_file` as a Path object, ensuring compatibility with `load_image`.
```
Comment (nicoddemus) [inline doc/en/how-to/types.rst:231]: I think this is also a bit redundant.

```suggestion
```

**Review by nicoddemus — APPROVED** [review]

**Comment by nicoddemus** [comment]
Thanks @mwychung!

Took the liberty of tweaking the page, the idea for the page is to be a reference for those that already know about the advantages of typing, and only are unsure how to use typing in the test suite itself.
